### PR TITLE
Store initial job jars on startup.

### PR DESCRIPTION
**Pull request by lossyrob**
This PR allows the user to supply a `spark.jobserver.job-jar-uris` configuration object that will cause SJS to store jars at the given URIs for the given application names. It behaves as if the first thing you do when SJS starts is post jars to the `localhost:8090/jars/appName` endpoint with a set of local jars.

The idea comes from @hectcastro.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -85,7 +86,7 @@ Then go ahead and start the job server using the instructions above.
 
 Let's upload the jar:
 
-    curl --data-binary @job-server-tests/target/job-server-tests-$VER.jar localhost:8090/jars/test
+    curl --data-binary @job-server-tests/target/scala-2.10/job-server-tests-$VER.jar localhost:8090/jars/test
```
Comment (lossyrob) [inline README.md]: This was required for my run of the test, is this a difference in `sbt` versions or a typo in the README?

Comment (velvia) [inline README.md]: @lossyrob thanks for the PR!   Yeah it's probably an oversight, a while back the path changed.

File: job-server/src/spark.jobserver/JobServer.scala [hunk 2]
```diff
@@ -66,6 +75,44 @@ object JobServer {
...
+
+      val initialJars =
+        initialJarsConfig
+          .asScala
+          .map { case (key, value) => (key, value.unwrapped.toString) }
+          .toMap
+
+      // Ensure that the jars exist
+      for(jarPath <- initialJars.values) {
+        val f = new java.io.File(jarPath)
```
Comment (velvia) [inline job-server/src/spark.jobserver/JobServer.scala]: So, the code here assumes that each of the strings in `job-jar-uris` is actually a local path, is that right?   I think this is an ok place to start, but maybe we should rename the config path as `job-jar-paths`?

File: job-server/src/main/resources/application.conf [hunk 3]
```diff
@@ -17,6 +17,12 @@ spark {
       rootdir = /tmp/spark-jobserver/filedao/data
     }
 
+    # To load up job jars on startup, place them here,
+    # with the app name as the key and the path to the jar as the value
+    # job-jar-uris {
+    #   test = ../job-server-tests/target/scala-2.10/job-server-tests_2.10-0.5.3-SNAPSHOT.jar
```
Comment (noorul) [inline job-server/src/main/resources/application.conf]: Can this be any URI? For example we store job jar in an HTTP server. Will this patch help us to put the url here and upload the jar?

File: job-server/src/spark.jobserver/JobServer.scala [hunk 4]
```diff
@@ -66,6 +75,44 @@ object JobServer {
...
+        if(!f.exists) {
+          val msg =
+            if(f.isAbsolute) {
+              s"Inital Jar File $jarPath does not exist"
+            } else {
+              s"Inital Jar File $jarPath (${f.getAbsolutePath}) does not exist"
+            }
+
+          throw new java.io.IOException(msg)
```
Comment (noorul) [inline job-server/src/spark.jobserver/JobServer.scala]: Does this means that it is all or none?

For exampl, say I have four apps and corresponding paths and somehow one of the path is failing. Does this mean none them will be uploaded?

File: job-server/src/spark.jobserver/JobServer.scala (hunk 2 above)
Comment (lossyrob) [inline job-server/src/spark.jobserver/JobServer.scala]: That makes sense!

File: job-server/src/main/resources/application.conf (hunk 3 above)
Comment (lossyrob) [inline job-server/src/main/resources/application.conf]: This currently only supports local paths. Adding support for other initial jar locations could be supported, though...is there a distinct need for that currently, or should we start with just local JARs?

File: job-server/src/spark.jobserver/JobServer.scala (hunk 4 above)
Comment (lossyrob) [inline job-server/src/spark.jobserver/JobServer.scala]: If there is an error with the initial JARs, then SJS won't start, and will inform you about the failing JAR. That behavior was by design, do you think another behavior would be better?

Comment (noorul) [inline job-server/src/spark.jobserver/JobServer.scala]: makes sense

File: job-server/src/spark.jobserver/JobServer.scala (hunk 2 above)
Comment (noorul) [inline job-server/src/spark.jobserver/JobServer.scala]: Do you think we should add test cases too?

File: job-server/src/spark.jobserver/JarManager.scala [hunk 5]
```diff
@@ -5,9 +5,12 @@ import spark.jobserver.io.JobDAO
 import spark.jobserver.util.JarUtils
 import org.joda.time.DateTime
 
+import java.nio.file.{Files, Paths}
+
 // Messages to JarManager actor
 case class StoreJar(appName: String, jarBytes: Array[Byte])
 case object ListJars
+case class StoreInitialJars(initialJars: Map[String, String])
```
Comment (velvia) [inline job-server/src/spark.jobserver/JarManager.scala]: Very minor nit: from JarManager's perspective it doesn't care if they are initial or not, could call the message `StoreJars`.

File: job-server/src/spark.jobserver/JarManager.scala [hunk 6]
```diff
@@ -18,16 +21,42 @@ case object JarStored
...
+              }
+            } catch {
+              case e: Exception =>
+                  logger.error(e.getMessage)
+                  false
+            }
+          }
+        }
+
+      sender ! (if(success) { JarStored } else { InvalidJar })
```
Comment (velvia) [inline job-server/src/spark.jobserver/JarManager.scala]: minor nit: space after if

File: job-server/src/spark.jobserver/JobServer.scala [hunk 7]
```diff
@@ -66,6 +75,45 @@ object JobServer {
...
+          .map { case (key, value) => (key, value.unwrapped.toString) }
+          .toMap
+
+      // Ensure that the jars exist
+      for(jarPath <- initialJars.values) {
+        val f = new java.io.File(jarPath)
+        if(!f.exists) {
+          val msg =
+            if(f.isAbsolute) {
+              s"Inital Jar File $jarPath does not exist"
```
Comment (velvia) [inline job-server/src/spark.jobserver/JobServer.scala]: spelling: Initial

File: job-server/src/spark.jobserver/JobServer.scala (hunk 2 above)
Comment (velvia) [inline job-server/src/spark.jobserver/JobServer.scala]: I think we can leave it, JobServer.scala is a bit hard to test since it's the outer app.

**Comment by velvia** [comment]
@lossyrob Thanks again;  except for very minor comments I think this is ready for merge.

File: job-server/src/spark.jobserver/JarManager.scala (hunk 5 above)
Comment (noorul) [inline job-server/src/spark.jobserver/JarManager.scala]: I agree, but I have one question, how do we differentiate both these classes. I think we should somehow document it.

File: job-server/src/spark.jobserver/JarManager.scala (hunk 6 above)
Comment (noorul) [inline job-server/src/spark.jobserver/JarManager.scala]: Don't we have this enabled in scalastyle?

File: job-server/src/spark.jobserver/JobServer.scala [hunk 8]
```diff
@@ -66,6 +75,45 @@ object JobServer {
...
+          .asScala
+          .map { case (key, value) => (key, value.unwrapped.toString) }
+          .toMap
+
+      // Ensure that the jars exist
+      for(jarPath <- initialJars.values) {
+        val f = new java.io.File(jarPath)
+        if(!f.exists) {
+          val msg =
+            if(f.isAbsolute) {
```
Comment (noorul) [inline job-server/src/spark.jobserver/JobServer.scala]: space after if?

File: job-server/src/spark.jobserver/JarManager.scala (hunk 5 above)
Comment (lossyrob) [inline job-server/src/spark.jobserver/JarManager.scala]: Makes sense, however I wanted to differentiate the difference between the `StoreJar` that took the `jarBytes` and this message, which actually provides the path to be read. Do you think this deserves differentiation in naming? If not I can just change to `StoreJars`.

I'll also add some doc strings on the messages, thanks @noorul

**Comment by lossyrob** [comment]
Sorry I've been out, I'll update this PR with the comments soon.

File: job-server/src/spark.jobserver/JarManager.scala (hunk 5 above)
Comment (lossyrob) [inline job-server/src/spark.jobserver/JarManager.scala]: How about `StoreLocalJars`?

Comment (velvia) [inline job-server/src/spark.jobserver/JarManager.scala]: @lossyrob `StoreLocalJars` sounds perfect!

**Comment by lossyrob** [comment]
This should be ready for merge now.

**Comment by noorul** [comment]
LGTM

**Comment by velvia** [comment]
@lossyrob Thanks again, everything looks good, going to merge.

**Pull request merged**